src/Mod/Sketcher/Gui/CommandSketcherTools.cpp

```diff
@@ -51,13 +51,92 @@
 #include "ViewProviderSketch.h"
 
 #include "DrawSketchHandlerOffset.h"
+#include "DrawSketchHandlerRotate.h"
 
 // Hint: this is to prevent to re-format big parts of the file. Remove it later again.
 // clang-format off
 using namespace std;
 using namespace SketcherGui;
 using namespace Sketcher;
 
+std::vector<int> getListOfSelectedGeoIds(bool forceInternalSelection)
+{
+    std::vector<int> listOfGeoIds = {};
+
+    // get the selection
+    std::vector<Gui::SelectionObject> selection;
+    selection = Gui::Selection().getSelectionEx(0, Sketcher::SketchObject::getClassTypeId());
+
+    // only one sketch with its subelements are allowed to be selected
+    if (selection.size() != 1) {
+        QMessageBox::warning(Gui::getMainWindow(),
+            QObject::tr("Wrong selection"),
+            QObject::tr("Select elements from a single sketch."));
+        return {};
+    }
+
+    // get the needed lists and objects
+    auto* Obj = static_cast<Sketcher::SketchObject*>(selection[0].getObject());
+    const std::vector<std::string>& subNames = selection[0].getSubNames();
+    if (!subNames.empty()) {
+
+        for (auto& name : subNames) {
+            // only handle non-external edges
+            if (name.size() > 4 && name.substr(0, 4) == "Edge") {
+                int geoId = std::atoi(name.substr(4, 4000).c_str()) - 1;
+                if (geoId >= 0) {
+                    listOfGeoIds.push_back(geoId);
+                }
+            }
+            else if (name.size() > 6 && name.substr(0, 6) == "Vertex") {
+                // only if it is a GeomPoint
+                int VtId = std::atoi(name.substr(6, 4000).c_str()) - 1;
+                int geoId;
+                Sketcher::PointPos PosId;
+                Obj->getGeoVertexIndex(VtId, geoId, PosId);
+                if (isPoint(*Obj->getGeometry(geoId))) {
+                    if (geoId >= 0) {
+                        listOfGeoIds.push_back(geoId);
+                    }
+                }
+            }
+        }
+    }
+
+    if (forceInternalSelection) {
+        for (auto geoId : listOfGeoIds) {
+            const Part::Geometry* geo = Obj->getGeometry(geoId);
+            if (isEllipse(*geo) || isArcOfEllipse(*geo) || isArcOfHyperbola(*geo) || isArcOfParabola(*geo) || isBSplineCurve(*geo)) {
+                const std::vector<Sketcher::Constraint*>& constraints = Obj->Constraints.getValues();
+                for (auto constr : constraints) {
+                    if (constr->Type == InternalAlignment && constr->Second == geoId) {
+                        if (std::find(listOfGeoIds.begin(), listOfGeoIds.end(), constr->First) == listOfGeoIds.end()) {
+                            // If the value is not found, add it to the vector
+                            listOfGeoIds.push_back(constr->First);
```

Comment:
I have marked (A) above.

So we are iterating vector ```listOfGeoIds```, this line here is adding new elements to it.

A for range construction as (A) uses iterators internally. When you add one element to the vector, the iterators are invalidated. Then you may be accessing another area (it is undefined behaviour). Here, geoId, on the second iteration of the loop (after adding the internal geometry to the array) gets value 22354.

If you abstract for a moment from the specific case, think that an vector is allocated a consecutive chunk of memory. When you add elements, the vector itself may need to grow. When it needs to grow, it may be reallocated elsewhere. Therefore, there is no guarantee your iterators continue to be valid.

Some implementations (I guess MSVC) may disregard the issue relying on no reallocation, which often may be true, but it is a serious issue.

The easiest way is to drop use of iterators. For example, remove the ```range for``` (A) and substitute it with a traditional indexed loop ```for( i;i<loopsize;i++)``. Here loopsize is the initial vector size (as you do not want to loop over the added internal geometry at the end of the array).

An alternative is to add the internalgeometry to another vector and after the loop merge them.

I am no allergic to traditional indexed loops when the use case is clear, so I would do the first. But if you want to do it the other way around, fine with me.

Reply:
Great thanks ! 
I will update asap